config/v1/types_infrastructure.go

```diff
@@ -505,9 +505,151 @@ type BareMetalPlatformStatus struct {
+	// speaker configuration for every failure domain hosting a control
+	// plane node. The list must contain at least one item.
+	//
+	// +kubebuilder:validation:MinItems:=1
+	// +listType=map
+	// +listMapKey=failureDomain
+	Speakers []OpenStackAPIBGPSpeaker `json:"speakers"`
+}
+
+// OpenStackFailureDomain specifies a failure domain for an OpenStack server.
+// Specifically, it specifies a set of values which will be set on all machines
+// using the failure domain.
+type OpenStackFailureDomain struct {
+	// name is an arbitrary, unique name for this failure domain. This name
+	// can be used to refer to this failure domain when building a BGP
+	// speaker configuration.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// computeZone specifies the OpenStack Compute availability zone for
+	// all servers in this failure domain.
+	//
+	// If not specified the servers are provisioned without reference to
+	// availability zones. Server placement is delegated to the OpenStack
+	// defaults.
+	// +optional
+	// + ---
+	// + TODO: define behaviour in case only one of computeZone and
+	// + storageZone is set.
+	ComputeZone string `json:"computeZone,omitempty"`
+
+	// storageZone specifies the OpenStack storage availability zone for
+	// all volumes in this failure domain.
+	//
+	// If not specified the volumes are provisioned without reference to
+	// availability zones. Volume placement is delegated to the OpenStack
+	// defaults.
+	// +optional
+	// + ---
+	// + TODO: define behaviour in case only one of computeZone and
+	// + storageZone is set.
+	StorageZone string `json:"storageZone,omitempty"`
+
+	// subnetID specifies an OpenStack subnet ID which will be attached as
+	// the first NIC of every server in this failure domain.
+	//
+	// +optional
+	SubnetID string `json:"subnetID,omitempty"`
+}
+
+// APILoadBalancerType defines how inbound traffic is routed to the API
+// servers.
+type APILoadBalancerType string
+
+const (
+	// APILoadBalancerTypeDefault lets the controller define a default
+	APILoadBalancerTypeDefault APILoadBalancerType = ""
+
+	// APILoadBalancerTypeVRRP uses VRRP to route inbound API traffic to
+	// the Control plane servers
+	APILoadBalancerTypeVRRP APILoadBalancerType = "VRRP"
+
+	// APILoadBalancerTypeBGP uses BGP to route inbound API traffic to the
+	// Control plane servers
+	APILoadBalancerTypeBGP APILoadBalancerType = "BGP"
+)
+
+// IsSet returns true if the field was filled in in configuration.
+func (lbT APILoadBalancerType) IsSet() bool {
+	return lbT != APILoadBalancerTypeDefault
+}
+
 // OpenStackPlatformSpec holds the desired state of the OpenStack infrastructure provider.
 // This only includes fields that can be modified in the cluster.
-type OpenStackPlatformSpec struct{}
+type OpenStackPlatformSpec struct {
+	// failureDomains is a list of failure domains available to Machines.
+	// Values specified in a failure domain will be set on Machines in that
+	// failure domain.
```

Comment:
What happens when there are no values in this list?

Reply:
addressed